plugins/optimization/src/main/java/org/qbicc/plugin/opt/ea/EscapeAnalysisOptimizeVisitor.java

```diff
@@ -0,0 +1,81 @@
+    private final CompilationContext ctxt;
+    private final NodeVisitor<Node.Copier, Value, Node, BasicBlock, ValueHandle> delegate;
+
+    public EscapeAnalysisOptimizeVisitor(final CompilationContext ctxt, final NodeVisitor<Node.Copier, Value, Node, BasicBlock, ValueHandle> delegate) {
+        this.ctxt = ctxt;
+        this.delegate = delegate;
+    }
+
+    public NodeVisitor<Node.Copier, Value, Node, BasicBlock, ValueHandle> getDelegateNodeVisitor() {
+        return delegate;
+    }
+
+    @Override
+    public Value visit(Node.Copier param, New original) {
+        final BasicBlockBuilder bbb = param.getBlockBuilder();
+        if (EscapeAnalysisState.get(ctxt).isNotEscapingMethod(original, bbb.getCurrentElement())) {
+            return stackAllocate(original.getClassObjectType(), bbb);
+        }
+
+        return NodeVisitor.Delegating.super.visit(param, original);
+    }
+
+    private Value stackAllocate(ClassObjectType type, BasicBlockBuilder bbb) {
+        // Copied from NoGcBasicBlockBuilder
+        Layout layout = Layout.get(ctxt);
+        Layout.LayoutInfo info = layout.getInstanceLayoutInfo(type.getDefinition());
+        CompoundType compoundType = info.getCompoundType();
+        LiteralFactory lf = ctxt.getLiteralFactory();
+        IntegerLiteral align = lf.literalOf(compoundType.getAlign());
+
+        // TODO David Lloyd: the alignment should be the minimum alignment of the lowered target type
+        // TODO David Lloyd: which we don't know until lower, but perhaps ObjectAccessLoweringBuilder could intercept that and set the minimum alignment to the minimum of the layout
+        Value ptrVal = bbb.stackAllocate(compoundType, lf.literalOf(1), align);
+
+        Value oop = bbb.valueConvert(ptrVal, type.getReference());
+        ValueHandle oopHandle = bbb.referenceHandle(oop);
+
+        // zero initialize the object's instance fields (but not the header fields that are defined in java.lang.Object)
+        LoadedTypeDefinition curClass = type.getDefinition().load();
+        while (curClass.hasSuperClass()) {
+            curClass.eachField(f -> {
+                if (!f.isStatic()) {
+                    bbb.store(bbb.instanceFieldOf(oopHandle, f), lf.zeroInitializerLiteralOfType(f.getType()), MemoryAtomicityMode.UNORDERED);
+                }
+            });
+            curClass = curClass.getSuperClass();
+        }
+
+        // now initialize the object header (aka fields of java.lang.Object)
+        initializeObjectHeader(oopHandle, layout, type.getDefinition().load().getType(), bbb);
+
+        return oop;
+    }
+
+    private void initializeObjectHeader(ValueHandle oopHandle, Layout layout, ObjectType objType, BasicBlockBuilder bbb) {
+        FieldElement typeId = layout.getObjectTypeIdField();
+        bbb.store(bbb.instanceFieldOf(oopHandle, typeId),  ctxt.getLiteralFactory().literalOfType(objType), MemoryAtomicityMode.UNORDERED);
```

Comment:
And there is related logic in BuildTimeHeap that knows about object headers too, since it has to generate the StructLiterals for the serialized objects.  This might be harder to factor into a common helper, but I agree with the comment that we need to have some way of finding all this code and keeping it in synch.

Reply:
Makes sense. I've created https://github.com/qbicc/qbicc/issues/638 to track the work to generalize the logic.